src/cocotb_tools/runner.py

```diff
@@ -427,13 +427,32 @@ def test(
         self.env["COCOTB_RESULTS_FILE"] = str(results_xml_file)
 
         cmds: Sequence[_Command] = self._test_command()
-        self._execute(cmds, cwd=self.test_dir)
+        simulator_exit_code: int = 0
+        try:
+            self._execute(cmds, cwd=self.test_dir)
+        except subprocess.CalledProcessError as e:
+            # It is possible for the simulator to fail but still leave results.
+            self.log.error("Simulation failed: %d", e.returncode)
+            simulator_exit_code = e.returncode
 
         # Only when running under pytest, check the results file here,
         # potentially raising an exception with failing testcases,
         # otherwise return the results file for later analysis.
         if pytest_current_test:
-            check_results_file(results_xml_file)
+            try:
+                (num_tests, num_failed) = get_results(results_xml_file)
+                if num_failed:
+                    self.log.error(
+                        "ERROR: Failed %d of %d tests.", num_failed, num_tests
+                    )
+                    error_code = 1 if simulator_exit_code == 0 else simulator_exit_code
+                    sys.exit(error_code)
```

Comment:
I was using it in two places at some point. I'll remove the intermediate variable.